File: Core/src/main/java/dev/lynith/core/ui/components/Screen.java
package dev.lynith.core.ui.components;

import dev.lynith.core.bridge.gui.IRenderer;
import dev.lynith.core.ui.components.features.Children;
import dev.lynith.core.ui.styles.ComponentStyles;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@Getter
public abstract class Screen extends Component<Screen, ComponentStyles<?, ?>> implements Children {

    private final List<Component<?, ?>> children = new ArrayList<>();

    @Override
    protected void postRender(IRenderer ctx, int mouseX, int mouseY, float delta) {
        for (Component<?, ?> child : children) {
            child.render(ctx, mouseX, mouseY, delta);
        }
        super.postRender(ctx, mouseX, mouseY, delta);
    }

    @Override
    protected void postInit() {
        super.postInit();
        for (Component<?, ?> child : children) {
            child.init();
        }
    }

    @Override
    public <T extends ComponentCallback> void on(Class<T> event, T callback, boolean once) {
        super.on(event, callback, once);
        for (Component<?, ?> child : children) {
            child.on(event, callback, once);
        }
    }

    @Override
    public ComponentStyles<?, ?> getStyles() {
        return null;
    }
}

File: Core/src/main/kotlin/dev/lynith/core/ui/components/Screen.kt
package dev.lynith.core.ui.components

import dev.lynith.core.Platform
import dev.lynith.core.bridge.gui.MCScreen
import dev.lynith.core.ui.BoundingBox
import dev.lynith.core.ui.callbacks.impl.*
import dev.lynith.core.ui.styles.ComponentStyles

abstract class Screen : ComponentWithChildren<Screen, ComponentStyles.BaseStyles<Screen>>() {
    var shouldPauseGame: Boolean = false
    var maxFramerate: Int = 60

    companion object {
        @JvmStatic
        var currentScreen: Screen? = null
    }

    override var bounds: BoundingBox = BoundingBox(
        width = Platform.renderer.windowWidth.toFloat(),
        height = Platform.renderer.windowHeight.toFloat()
    )

    override fun preInit() {
        this.screen = this
        currentScreen = this

        once<Destroyed> {
            unregister()
            currentScreen = null
        }
    }

    override fun render(mouseX: Int, mouseY: Int, delta: Float) {
        // Empty
    }

    override var styles = ComponentStyles.BaseStyles<Screen>()

    override fun child(child: Component<*, *>) {
        super.child(child)
        child.screen = this
    }

    // Utils

    open fun toMCScreen(): MCScreen {
        return WrappedScreen(this)
    }

    protected open class WrappedScreen(
        private val screen: Screen
    ) : MCScreen() {
        var prevX: Int? = null
        var prevY: Int? = null

        override fun render(mouseX: Int, mouseY: Int, delta: Float) {
            super.render(mouseX, mouseY, delta)

            val mouseXScaled = mouseX * Platform.renderer.scaleFactor
            val mouseYScaled = mouseY * Platform.renderer.scaleFactor

            if (prevX != mouseXScaled || prevY != mouseYScaled) {
                prevX = mouseXScaled
                prevY = mouseYScaled
                screen.emit(CursorMoved(prevX!!, prevY!!))
            }

            screen.wrappedRender(mouseXScaled, mouseYScaled, delta)
        }

        override fun init() {
            super.init()
            screen.wrappedInit()
        }

        override fun mouseClicked(mouseX: Int, mouseY: Int, button: Int) {
            super.mouseClicked(mouseX, mouseY, button)
            screen.emit(Pressed(mouseX, mouseY, button))
        }

        override fun mouseReleased(mouseX: Int, mouseY: Int, button: Int) {
            super.mouseReleased(mouseX, mouseY, button)
            screen.emit(Released(mouseX, mouseY, button))
        }

        override fun shouldPauseGame(): Boolean {
            return screen.shouldPauseGame
        }

        override fun resized() {
            super.resized()

            screen.bounds = BoundingBox(
                width = Platform.renderer.windowWidth.toFloat(),
                height = Platform.renderer.windowHeight.toFloat()
            )

            screen.emit(WindowResized(screen.bounds.width, screen.bounds.height))
        }

        override fun closed() {
            super.closed()
            screen.emit(Destroyed())
        }
    }
}
